pkg/detectors/detectors.go

```diff
@@ -39,14 +39,24 @@ type EndpointCustomizer interface {
 	DefaultEndpoint() string
 }
 
+// Result is a secret that has been found by a Detector and optionally verified.
+// It contains the raw secret identification data, redacted data, and extra data.
 type Result struct {
 	// DetectorType is the type of Detector.
 	DetectorType detectorspb.DetectorType
 	// DetectorName is the name of the Detector. Used for custom detectors.
 	DetectorName string
 	// DecoderType is the type of Decoder.
 	DecoderType detectorspb.DecoderType
-	Verified    bool
+	// Verified is true if the secret has been confirmed to be a valid secret.
+	Verified bool
+	// HasVerificationIndeterminate is true if the secret could NOT be verified.
+	// This is used to differentiate between secrets that are NOT verified and secrets that COULD NOT be verified.
+	// Ex: A 429 response from a service, or a timeout would set this to true.
+	HasVerificationIndeterminate bool
+	// VerificationIndeterminateError holds the error if the secret could NOT be verified.
+	// This is used specifically for cases that resulted in indeterminate resolution, such as a 429 or timeout.
+	VerificationIndeterminateError error
```

Comment:
Thoughts on combining these two attributes into just the `error`, where `nil` indicates `HasVerificationIndeterminate == false`? Or rather, is `(true, nil)` / `(false, non-nil)` valid?

Just throwing it out there for opinions. It might be more confusing tying it together.

Reply:
So my thoughts/push back on that are the following:
1.The dual fields HasVerificationIndeterminate and VerificationIndeterminateError very clearly separate the concern of whether an indeterminate verification state was reached and the specific error that occurred in that case. By collapsing these into one field, i feel like you lose some of that semantic clarity.
2.A nil error is universally understood in Go as "no error occurred". Using nil to signify false for HasVerificationIndeterminate could be counter-intuitive, as this could be interpreted as "the verification was indeterminate, but no error occurred", which may not be the intended meaning. It also feels a bit weird to me, but that's just me.
3. Again me likely being annoying, but the nil check is more verbose,If you use the VerificationIndeterminateError field to signify both the occurrence of an error and the indeterminate verification state, your code will frequently need to check if result.VerificationIndeterminateError != nil to determine whether verification was indeterminate.
This is more verbose and less direct than checking if result.HasVerificationIndeterminate, which clearly and succinctly communicates the intent of the check.
4.Using an error object to store a boolean state can be seen as a misuse of the error type. Errors in Go are generally used to hold information about something that went wrong in the execution of the program, not to hold state information. Using the error in this way in my opinion makes the code less clear, as future readers of the code might expect VerificationIndeterminateError to be used in the conventional way, i.e., to hold error information, not state information.

That being said, I think it makes sense to get additional opinions.

Reply:
Thanks for the well thought out response to my half baked question :)

Yeah, I like the explicitness of the two separate variables. My original thought was along the lines of something like an `Option` enum where a non-nil error indicates `Some(error)` and a nil error indicates `None`.

It semantically represents two states, rather than four when separated into `bool` and `error`. What would it mean if `VerificationIndeterminateError` is `non-nil` and `HasVerificationIndeterminate` is `false`? Or vice versa?

Reply:
Ahh I see what you mean. Also the alternative also would be a weird state. `HasVerificationIndetermindate` is true and non-nil error.

So a few ideas come to mind:
1. We can simply go off the basis that if `HasVerification...` is not set, we don't care about the error, and vice versa. (i don't like this that much)
2. Make both the fields unexported and use helpers to get and set:
ex:
```
func (r Result) GetVerificationIndeterminate() (bool, error) {
    return r.hasVerificationIndeterminate, r.verificationIndeterminateError
}

func (r Result) SetVerificationIndeterminate(hasIndeterminate bool, err error) {
    if hasIndeterminate && err != nil {
        r.hasVerificationIndeterminate = true
        r.verificationIndeterminateError = err
    } else if !hasIndeterminate && err == nil {
        r.hasVerificationIndeterminate = false
        r.verificationIndeterminateError = nil
    } else {
        panic("Invalid state")
    }
}
```
3. something akin to state enumeration, (maybe what you were referring to?):
ex:
```
type VerificationState int

const (
    VerificationUnknown VerificationState = iota
    VerificationConfirmed
    VerificationIndeterminate
)

func (r Result) GetVerificationState() VerificationState {
    return r.verificationState
}

func (r Result) GetVerificationIndeterminateError() error {
    if r.verificationState != VerificationIndeterminate {
        return nil
    }
    return r.verificationIndeterminateError
}

func (r Result) SetVerificationState(state VerificationState, err error) {
    switch state {
    case VerificationUnknown, VerificationConfirmed:
        if err != nil {
            ...
        }
        r.verificationState = state
        r.verificationIndeterminateError = nil
    case VerificationIndeterminate:
        if err == nil {
            ...
        }
        r.verificationState = state
        r.verificationIndeterminateError = err
    default:
        ...
    }
}
```

Reply:
Hm.. none of these options seem great and I'm probably over-complicating the design with this theoretical..

Reply:
What if `HasVerificationIndeterminate` just became a method that does the nil check?